nucypher/network/nodes.py

```diff
@@ -163,13 +163,13 @@ class Learner:
     __DEFAULT_MIDDLEWARE_CLASS = RestMiddleware
 
     LEARNER_VERSION = LEARNING_LOOP_VERSION
+    LOWEST_COMPATIBLE_VERSION = 2   # Disallow versions lower than this
```

Comment:
Feels like a semantic version tuple would be more intuitively understandable here.

Reply:
YES! I though something like that but also had second thoughts regarding how to encode this efficiently.

Reply:
This raises two questions from me:

 - What does a minor vs major bump mean?
 - How and when does the version get updated?

Reply:
Both good questions. The current design in `main` implicitly assumes all versions are going to be back-wards compatible, in the sense that if you detect a node from a past version, no special handling of the metadata is performed. There are cases, like our particular bump from version 1 to version 2, where this is not desirable (as this is a "major" change in semver parlance), but I certainly can imagine other cases where we bump the version and at the same time try to maintain compatibility with the old version (a "minor" change). The approach taken in this PR is very simplistic, and I think there's room for improvement, but at least ensures that no metadata from version 1 will linger on the network.

Reply:
:thinking:  Interesting - I suggest we open an Issue to track and leave it for a future PR

Reply:
> What does a minor vs major bump mean?

Major bump == `LOWEST_COMPATIBLE_VERSION` bump
Minor bump == `LEARNER_VERSION` bump
That's essentially what these variables are. Arranging them in a tuple just makes it more obvious.

Now I must say that based on the logic of `Ursula.from_bytes()` it seems that the treatment of versions is more restricted in some ways and more relaxed in others compared to the semver default. What happens now:
- if the version of the serialized data is within a certain range, expect it to conform to the format *exactly* 
- if the version of the serialized data is too far in the past, or higher than the class version, try to decode parts of it anyway (since it's only used for the error message, I guess it's fine)

That raises the question, if `payload_splitter()` can decode something encoded by the code with a different `LEARNER_VERSION` without skipping anything, why was `LEARNER_VERSION` bumped in the first place? 

How I would expect it to work:
- If the major version differs, raise an error. Possibly still try to extract some info to show in the error message.
- If the minor version of the serialized data is higher, load all the data you are aware of and skip the rest
- If the minor version of the serialized data is lower, load all the available data and fill the missing data with defaults.

This assumes that when the minor version is bumped, a new field is appended to the serialized data, and nothing else is changed. This will also require keeping a "history" of added fields vs minor versions so that the deserialization of data from the past could be performed safely. Not sure if you want to go there.

Reply:
The way I am planning on handling this sort of thing in the VersionedBytestrings PR is by using the BytestringKwargifier, which takes the class to deserialize as an input, we can return a different subclass for each version.

So, if Ursula has some changes, we just need to keep the old V1 Ursula accessible and then return that class for any Ursula bytestring that has V1 metadata.

Does this makes sense?

Reply:
How can we keep a whole other `Ursula` class accessible? What's the mechanism here?

Reply:
Kind of like what's happening in this test:
https://github.com/nucypher/bytestringSplitter/blob/master/tests/test_kwargifiers.py#L127

Reply:
What I meant is that you'll have to have the V1 Ursula class defined somewhere. Which is huge, and has links to other parts of the codebase.

Reply:
Well yeah... it depends on where the future takes us.  If the V1 can be the super class of the V2, then it's not so bad, we just change/add some methods.

If it's a total brand new distinct class, then it's less fun.

Reply:
I just realized, something similar to this situation is already happening in the case of TreasureMaps vs SignedTreasureMaps.

Two distinct classes with different bytestring de/serializations, one a subclass of the other.

[here](https://github.com/nucypher/nucypher/blob/bc3cd71e3047a7b0b50b9a8d5a4c71c1a0d34107/nucypher/policy/collections.py)